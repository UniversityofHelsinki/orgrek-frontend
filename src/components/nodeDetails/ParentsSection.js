import React from 'react';
import EditableAccordion from '../EditableAccordion';
import { useTranslation } from 'react-i18next';
import HierarchyTable from '../attributes/HierarchyTable';
import useContentLanguage from '../../hooks/useContentLanguage';
import Placeholder from '../Placeholder';
import { authActions } from '../../auth';
import { useNodeId } from '../../hooks/useNodeId';
import EditableContent from '../EditableContent';
import ParentEditor from './ParentEditor';
import { useSaveParentsMutation } from '../../store';
import { useParents } from '../../hooks/useParents';
import useFilterParentsByDate from '../../hooks/useFilterParentsByDate';

const ParentsSection = () => {
  const { t } = useTranslation();
  const { parents, isFetching } = useParents();
  const currentNodeId = useNodeId();
  const contentLanguage = useContentLanguage();

  const data = parents[contentLanguage] || [];
  const title = t('upper_units');
  const empty = data.length === 0;
  let parentUniqueId = -1;
  const [saveParents] = useSaveParentsMutation();

  const filteredByDateParents = useFilterParentsByDate(data);

  const asAttribute = (nodeId) => (hierarchy) => ({
    id: hierarchy.edgeId,
    nodeId: nodeId,
    value: hierarchy.hierarchy,
    startDate: hierarchy.startDate,
    endDate: hierarchy.endDate,
    isNew: Boolean(hierarchy.isNew),
    deleted: Boolean(hierarchy.deleted),
  });

  const asFormValues = data.reduce((a, c) => {
    parentUniqueId = c.uniqueId;
    a[c.uniqueId] = c.hierarchies.map(asAttribute(c.id));
    return a;
  }, {});

  const asEdge = (parent) => ({
    id: parent.id,
    hierarchy: parent.value,
    childUniqueId: currentNodeId,
    parentNodeId: parent.nodeId,
    startDate: parent.startDate,
    endDate: parent.endDate,
    isNew: parent.isNew,
    deleted: parent.deleted,
  });

  const handleSubmit = (parents) => {
    //const all = Object.values(parents).flat();
    //const edges = all.map(asEdge);
    const edges = Object.entries(parents)
      .map(([key, values]) =>
        values.map((value) => ({ ...asEdge(value), parentNodeId: key }))
      )
      .flat();
    return saveParents({ edges, nodeId: currentNodeId }).unwrap();
  };

  if (isFetching) {
    return <EditableAccordion title={title} loading />;
  }

  return (
    <EditableAccordion title={title}>
      <EditableContent
        editorComponent={<ParentEditor />}
        initialValues={asFormValues}
        onSubmit={handleSubmit}
        successMessage={t('parentInfo.saveSuccess')}
        errorMessage={t('parentInfo.saveError')}
        authActions={authActions.parents}
      >
        <Placeholder empty={empty} placeholder={t('upperUnits.empty')}>
          <HierarchyTable data={filteredByDateParents} summary={title} />
        </Placeholder>
      </EditableContent>
    </EditableAccordion>
  );
};

export default ParentsSection;
