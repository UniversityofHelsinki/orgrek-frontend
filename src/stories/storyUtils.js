export const waitForAnimations = async () =>
  new Promise((resolve) => setTimeout(resolve, 200));
