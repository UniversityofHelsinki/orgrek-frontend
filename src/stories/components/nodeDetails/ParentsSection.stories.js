import ParentsSection from '../../../components/nodeDetails/ParentsSection';
import {
  mockGetParents,
  mockSaveParents,
  withMockStore,
} from '../../../mockStore';
import { userEvent, waitFor, within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { waitForAnimations } from '../../storyUtils';

// Use a fixed date to ensure that tests always have a consistent result
const now = new Date('2023-03-22T14:28:00+0200');

const nodeId = '1';
const selectedHierarchy = 'talous,toiminnanohjaus';

export default {
  component: ParentsSection,
  parameters: {
    systemTime: now,
    reactRouter: {
      searchParams: {
        uid: nodeId,
      },
    },
  },
  decorators: [
    withMockStore({
      tree: {
        selectedHierarchy,
      },
      dr: {
        selectedDay: now,
      },
    }),
  ],
};

export const Default = {
  parameters: {
    msw: {
      handlers: [
        mockGetParents(nodeId, now, selectedHierarchy, {
          fi: [
            {
              id: '1000',
              edgeId: 20001,
              uniqueId: 1000000,
              startDate: '2009-01-01',
              endDate: null,
              hierarchies: [
                {
                  hierarchy: 'talous',
                  startDate: '2000-01-01',
                  endDate: null,
                  edgeId: 20003,
                },
                {
                  hierarchy: 'toiminnanohjaus',
                  startDate: '2000-01-01',
                  endDate: null,
                  edgeId: 20004,
                },
              ],
              fullName: 'Parent node 1',
              language: 'fi',
            },
            {
              id: '1001',
              edgeId: 20002,
              uniqueId: 1000001,
              startDate: null,
              endDate: null,
              hierarchies: [
                {
                  hierarchy: 'talous',
                  startDate: '2000-01-01',
                  endDate: null,
                  edgeId: 20004,
                },
                {
                  hierarchy: 'toiminnanohjaus',
                  startDate: '2000-01-01',
                  endDate: null,
                  edgeId: 20005,
                },
              ],
              fullName: 'Parent node 2',
              language: 'fi',
            },
          ],
        }),
        mockSaveParents(),
      ],
    },
  },
  decorators: [],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('Yläyksiköt')).toBeInTheDocument();
    });
    await expect(canvas.getByText('Parent node 1')).toBeInTheDocument();
    await expect(canvas.getByText('Parent node 2')).toBeInTheDocument();
  },
};

export const Empty = {
  parameters: {
    msw: {
      handlers: [
        mockGetParents(nodeId, now, selectedHierarchy, {
          fi: [
            {
              id: '1000',
              edgeId: 20001,
              uniqueId: 1000000,
              startDate: '2009-01-01',
              endDate: null,
              hierarchies: [
                {
                  hierarchy: 'talous',
                  startDate: '2000-01-01',
                  endDate: '2009-12-31',
                  edgeId: 20003,
                },
              ],
              fullName: 'Parent node 1',
              language: 'fi',
            },
          ],
        }),
        mockSaveParents(),
      ],
    },
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('Yläyksiköt')).toBeInTheDocument();
    });

    await expect(canvas.queryByText('Muokkaa')).not.toBeVisible();

    await expect(canvas.getByRole('button')).toHaveAttribute(
      'aria-expanded',
      'false'
    );
  },
};

export const EditMode = {
  ...Default,
  play: async (context) => {
    await Default.play(context);

    const canvas = within(context.canvasElement);

    await userEvent.click(canvas.getByText('Muokkaa'));

    await waitFor(async () => {
      await expect(canvas.getByText('Tallenna')).toBeDisabled();
    });

    await expect(canvas.getByText('Parent node 1')).toBeInTheDocument();
    await expect(canvas.getByText('Parent node 2')).toBeInTheDocument();
  },
};

export const Modified = {
  ...EditMode,
  play: async (context) => {
    await EditMode.play(context);

    const canvas = within(context.canvasElement.parentElement);

    await userEvent.click(canvas.getAllByLabelText(/Hierarkia/)[0]);
    await userEvent.click(canvas.getAllByRole('option')[1]);
    await userEvent.type(
      canvas.getAllByLabelText('Voimassaolo päättyy')[0],
      '31.12.2023',
      { delay: 100 }
    );

    await waitFor(async () => {
      await expect(canvas.getByText('Tallenna')).toBeEnabled();
    });
  },
};

export const DeletedRow = {
  ...EditMode,
  play: async (context) => {
    await EditMode.play(context);

    const canvas = within(context.canvasElement.parentElement);

    await userEvent.click(canvas.getAllByLabelText(/Toiminnot/)[1]);
    await userEvent.click(canvas.getByText('Poista rivi'));
    await expect(
      canvas.getByText(
        'Poistettu: Viralliset yksiköt, voimassa 1.1.2000 alkaen'
      )
    ).toBeInTheDocument();

    // Deleted row appears with an animation, so wait for it before running
    // a11y tests
    await waitForAnimations();
  },
};
