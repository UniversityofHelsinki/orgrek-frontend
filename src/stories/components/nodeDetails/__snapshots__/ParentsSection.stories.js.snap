// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/nodeDetails/ParentsSection Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Yläyksiköt",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Muokkaa",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Ei yläyksiköitä",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-parentssection--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/ParentsSection Empty smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "Yläyksiköt",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Muokkaa",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Ei yläyksiköitä",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-parentssection--empty",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
